pc2ccs/pc2v9
# I564 support per problem output limit

### Description of what the PR does

Adds support for setting "max output limit" on a per-problem basis (whereas currently, all problems acquire the same "global" output limit, either the system default or a value specified using the `max-output-size-K` keyword in the `contest.yaml` file.)

Specifically, adds a "max output in KB" field to the `Problem` class, and provides both interactive  (via the `EditProblemPane` GUI) and YAML (via the ContestSnakeYAMLLoader class) support for setting the max output value for a problem, overriding any value specified in `contest.yaml`.

### Issue which the PR fixes

Fixes #564 

### Environment in which the PR was developed (OS,IDE, Java version, etc.)

Windows 10.1, Java 1.8.0_201, Eclipse Version: 2019-12 (4.14.0)

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)

#### Test 1:  JUnits
- Run updated JUnit `/pc2v9/test/edu/csus/ecs/pc2/core/model/ProblemTest`, which has updates to test that problems get the correct default max output size if none is specified, and that setting an explicit per-problem output limit works.
- Run updated JUnit `/pc2v9/test/edu/csus/ecs/pc2/imports/ccs/ContestSnakeYAMLLoaderTest`, which has updates to test various combinations of output limit specifications (default, limit in contest.yaml but not in problem.yaml, limit in problem.yaml, etc.)
- Run the complete JUnit test suite; verify that all JUnits pass.

NOTE:  all the following tests assume that the contest which is loaded into the server has (at least) one problem configured.

#### Test 2: Default max output
- Start a server which loads a contest where there are no specifiers for max output size in either contest.yaml or any problem.yaml.
- Start an Admin, select the ConfigureContest>Problems tab, and verify that every problem has the default MaxOutput(KB) value (512)

#### Test 3: max output defined by `contest.yaml`
- Start a server which loads a contest where there is a `max-output-size-K` setting in `contest.yaml`, but NO settings of max output value in any problem.yaml.
- Verify that each problem  has the Max Output(KB) value specified in the `contest.yaml` file.

#### Test 4: Max output in `problem.yaml` overrides default
- Start a server which loads a contest where there is NO `max-output-size-K` setting in `contest.yaml`, but there IS such a setting in a `problem.yaml` file.
- Verify that the problem containing the max-output-size-K setting has the Max Output(KB) value specified in the `problem.yaml` file.

#### Test 5: Max output in `problem.yaml` overrides `contest.yaml` setting
- Start a server which loads a contest where there is a `max-output-size-K` setting in `contest.yaml`, and ALSO a (different) setting of max output value in a problem.yaml.
- Verify that the problem containing the max-output-size-K setting has the Max Output(KB) value specified in the `problem.yaml` file.

#### Test 6: Max output in "limits" section of `problem.yaml` is recognized
- Start a server which loads a contest where there is a `max-output-size-K` setting in `contest.yaml`, and where a `problem.yaml` file has a `limits:` section containing an `output:` keyword specifying a unique max output value (different from the 512K default and also different from that specifified in `contest.yaml`).
- Verify that the problem containing the `limits: output:` setting has the Max Output(KB) value specified in the limits: section of the `problem.yaml` file.

#### Test 7: Interactive setting of problem output limit
- Start a server which loads a contest with at least one problem.
- On the Admin "Problems" pane, select a problem and click "Edit".
- Enter a non-zero value into the "Max output size (KB) textbox.
- Click "Update".
- Exit the Admin and the Server.
- Restart the server and Admin.
- Verify that the edited problem retained the new max output value.

#### Test 8: Interactive switch to global/default max output limit
- Start a server which loads a contest where there is a `max-output-size-K` setting in `contest.yaml`,
- On the Admin "Problems" pane, select a problem and click "Edit".
- Enter the value zero into the "Max output size (KB)" textbox.  (This is how you specify "no per-problem output limit; use the global setting").
- Click "Update".
- Exit the Admin and the Server.
- Restart the server and Admin.
- Verify that the edited problem retained the max output value defined in the `contest.yaml` file.
